### PR TITLE
Support cuDNN frontend scaled dot product attention for FP8. Part- 1(forward)

### PR DESCRIPTION
**Pull request by wenscarl**
Add cudnn frontend support of scaled dot product attention for FP8 forward. doc [here](https://github.com/NVIDIA/cudnn-frontend/blob/98ca4e1941fe3263f128f74f10063a3ea35c7019/docs/operations/Attention.md).

NOTE: this feature relies on cudnn-frontend v1.6.1 which is not in XLA yet.

### CODE REVIEW

File: xla/service/gpu/cublas_cudnn.cc [hunk 1]
```diff
@@ -75,6 +75,8 @@ const absl::string_view kCudnnConvReorderFilterAndBiasCallTarget =
 const absl::string_view kCudnnNormCallTarget = "__cudnn$norm";
 
 // fMHA forward call targets.
+const absl::string_view kCudnnfMHASoftmaxF8CallTarget =
+    "__cudnn$fmhaSoftmax$f8";
```
Comment (philipphack) [inline xla/service/gpu/cublas_cudnn.cc:79]: Should this be __cudnn$fmhaSoftmaxF8?

File: xla/service/gpu/cublas_cudnn.cc [hunk 2]
```diff
@@ -157,6 +169,10 @@ bool IsCustomCallTofMHA(const HloInstruction& hlo) {
   return (IsFwdCustomCallTofMHA(hlo) || IsBwdCustomCallTofMHA(hlo));
 }
 
+bool IsCustomCallTofMHAF8(const HloInstruction& hlo) {
+  return (IsFwdCustomCallTofMHAF8(hlo));  // TODO(shuw)
```
Comment (philipphack) [inline xla/service/gpu/cublas_cudnn.cc:173]: Remove the parentheses. I'm guessing this will also cover the backward case?

File: xla/service/gpu/cublas_cudnn.h [hunk 3]
```diff
@@ -67,6 +67,8 @@ enum class CudnnfMHAKind {
   kBackwardSoftmax,
   kBackwardScaleBiasSoftmax,
   kBackwardScaleBiasSoftmaxDropout,
+  kSoftmaxf8,
```
Comment (philipphack) [inline xla/service/gpu/cublas_cudnn.h:70]: I'd change this to kSoftmaxF8 and kBackwardSoftmaxF8.

File: xla/service/gpu/cudnn_workspace_rewriter.cc [hunk 4]
```diff
@@ -225,6 +225,56 @@ absl::StatusOr<se::gpu::CudnnGraph> HloCustomCallToCuDnnGraph(
...
+    TF_ASSIGN_OR_RETURN(const xla::gpu::CudnnfMHAKind kind,
+                        xla::gpu::GetCudnnfMHAKind(custom_call));
+
+    TF_ASSIGN_OR_RETURN(
+        const auto gpu_config,
+        custom_call->backend_config<xla::gpu::GpuBackendConfig>());
+    const xla::gpu::CudnnfMHABackendConfig& config =
+        gpu_config.cudnn_fmha_backend_config();
+    Shape intermediate_tensor_shape(config.intermediate_tensor_shape());
+    absl::InlinedVector<Shape, 4> output_shapes = {
```
Comment (philipphack) [inline xla/service/gpu/cudnn_workspace_rewriter.cc:238]: Why does this have size 4?

File: xla/service/gpu/gpu_fused_mha_runner.cc [hunk 5]
```diff
@@ -83,6 +83,90 @@ absl::Status RunFusedMHA(GpufMHAParams params, se::Stream *stream,
...
+          ? se::DeviceMemory<OutputType>(*params.activation_buffer)
+          : se::DeviceMemoryBase();
+
+  se::dnn::AlgorithmDesc algorithm = params.config->algorithm;
+  if (options.runner_cache) {
+    algorithm = options.runner_cache->ToAlgorithmDesc();
+  }
+
+  absl::Status run_status = absl::OkStatus();
+  switch (params.config->kind) {
```
Comment (philipphack) [inline xla/service/gpu/gpu_fused_mha_runner.cc:145]: Do you need a switch here? Even with backward it's only two cases.

File: xla/service/gpu/gpu_fused_mha_runner.cc [hunk 6]
```diff
@@ -83,6 +83,90 @@ absl::Status RunFusedMHA(GpufMHAParams params, se::Stream *stream,
...
+          params, stream, options, lhs_bmm1_buffer, rhs_bmm1_buffer,
+          rhs_bmm2_buffer, descale_q_buffer, descale_k_buffer, descale_v_buffer,
+          descale_s_buffer, scale_s_buffer, scale_o_buffer, amax_s_buffer,
+          amax_o_buffer, output_buffer, scratch_memory, activation_buffer);
+      break;
+    default:
+      return Internal("Invalid cuDNN fMHA f8 kind");
+  }
+
+  if (!run_status.ok()) {
```
Comment (philipphack) [inline xla/service/gpu/gpu_fused_mha_runner.cc:157]: Why not use TF_RETURN_IF_ERROR on RunFusedMHAF8?

File: xla/service/gpu/gpu_fused_mha_runner.cc [hunk 7]
```diff
@@ -524,6 +671,35 @@ GpufMHAConfig::AsDnnFusedMHAOpConfig() const {
   return config;
 }
 
+absl::StatusOr<se::dnn::FusedMHAOp::Config>
```
Comment (philipphack) [inline xla/service/gpu/gpu_fused_mha_runner.cc:674]: Why was this moved?

File: xla/service/gpu/gpu_fused_mha_runner.cc [hunk 8]
```diff
@@ -608,6 +812,40 @@ GpufMHABackwardConfig::AsDnnFusedMHABackwardOpConfig() const {
...
+      return RunGpuFMHAF8Impl<tsl::float8_e4m3fn, tsl::float8_e4m3fn>(
+          params, stream, scratch_buffer, options);
+    case F8E5M2:
+      return RunGpuFMHAF8Impl<tsl::float8_e5m2, tsl::float8_e5m2>(
+          params, stream, scratch_buffer, options);
+    default:
+      return absl::UnimplementedError(absl::StrFormat(
+          "Unimplemented fused MHA with %s", ToString(fmha_config)));
+  }
+  return absl::OkStatus();
```
Comment (philipphack) [inline xla/service/gpu/gpu_fused_mha_runner.cc:846]: This line can't be reached.

File: xla/service/gpu/gpu_fused_mha_runner.cc [hunk 9]
```diff
@@ -715,5 +953,26 @@ std::string ToString(const GpufMHAConfig &config) {
   return result;
 }
 
+// std::string ToString(const GpufMHAConfig &config) {
```
Comment (philipphack) [inline xla/service/gpu/gpu_fused_mha_runner.cc:956]: Remove this comment.

File: xla/service/gpu/gpu_fused_mha_runner.h [hunk 10]
```diff
@@ -103,12 +117,39 @@ struct GpufMHABackwardDescriptor {
   bool force_deterministic;
 };
 
+// // Structure to describe static properties of a GPU fused Multi-Headed
```
Comment (philipphack) [inline xla/service/gpu/gpu_fused_mha_runner.h:120]: Remove.

File: xla/service/gpu/gpu_fused_mha_runner.h [hunk 11]
```diff
@@ -227,6 +299,74 @@ struct GpufMHABackwardParams {
   std::optional<se::DeviceMemoryBase> seqlen_k_buffer;
 };
 
+class FusedMultiHeadedAttentionF8Runner {
+ public:
+  using Repr = std::variant<
+      std::monostate,  // To allow XXX default ctor
```
Comment (philipphack) [inline xla/service/gpu/gpu_fused_mha_runner.h:305]: Revise the comment.

File: xla/service/gpu/gpu_fused_mha_runner.h [hunk 12]
```diff
@@ -227,6 +299,74 @@ struct GpufMHABackwardParams {
...
+          std::unique_ptr<se::dnn::LazyOpRunner<se::dnn::FusedMHAF8Op>>>(
+        repr_));
+    return std::get<
+               std::unique_ptr<se::dnn::LazyOpRunner<se::dnn::FusedMHAF8Op>>>(
+               repr_)
+        .get();
+  }
+
+ private:
+  //  The CreateRunner function is defined as static because it
```
Comment (philipphack) [inline xla/service/gpu/gpu_fused_mha_runner.h:340]: Is this comment needed?

File: xla/service/gpu/gpu_fused_mha_runner.h [hunk 13]
```diff
@@ -425,6 +589,7 @@ absl::Status RunGpuFMHABackward(
     RunFusedMHABackwardOptions = {});
 
 std::string ToString(const GpufMHAConfig& config);
+// std::string ToString(const GpufMHAF8Config& config);
```
Comment (philipphack) [inline xla/service/gpu/gpu_fused_mha_runner.h:592]: Remove.

File: xla/service/gpu/ir_emitter_unnested.cc [hunk 14]
```diff
@@ -1177,6 +1176,91 @@ absl::Status IrEmitterUnnested::EmitFusedMHABackwardThunk(
...
+  const HloInstruction* lhs_bmm1 = instr->operand(0);
+  const HloInstruction* rhs_bmm1 = instr->operand(1);
+  const HloInstruction* rhs_bmm2 = instr->operand(2);
+  const HloInstruction* descale_q = instr->operand(3);
+  const HloInstruction* descale_k = instr->operand(4);
+  const HloInstruction* descale_v = instr->operand(5);
+  const HloInstruction* descale_s = instr->operand(6);
+  const HloInstruction* scale_s = instr->operand(7);
+  const HloInstruction* scale_o = instr->operand(8);
+  TF_ASSIGN_OR_RETURN(BufferAllocation::Slice lhs_bmm1_slice,
```
Comment (philipphack) [inline xla/service/gpu/ir_emitter_unnested.cc:1190]: Could introduce a small lambda to make this a bit shorter.

File: xla/service/gpu/ir_emitter_unnested.cc [hunk 15]
```diff
@@ -1177,6 +1176,91 @@ absl::Status IrEmitterUnnested::EmitFusedMHABackwardThunk(
...
+  bool has_activation = xla::ShapeUtil::TupleElementCount(instr->shape()) == 5;
+  if (has_activation) {
+    TF_ASSIGN_OR_RETURN(activation_slice, GetAllocationSliceForHlo(instr, {3}));
+  }
+
+  TF_ASSIGN_OR_RETURN(const xla::gpu::CudnnfMHAKind kind,
+                      xla::gpu::GetCudnnfMHAKind(instr));
+  assert(kind == CudnnfMHAKind::kScaleSoftmaxf8);
+  // BufferAllocation::Slice mask_slice;
+  // std::optional<Shape> mask_shape;
```
Comment (philipphack) [inline xla/service/gpu/ir_emitter_unnested.cc:1227]: Remove.

File: xla/service/gpu/tests/gpu_fused_mha_test.cc [hunk 16]
```diff
@@ -1089,6 +1089,119 @@ class FlashAttentionBMMScalePaddingMaskSoftmaxBMM
   }
 };
 
+class FlashAttentionBMMScalePaddingMaskSoftmaxBMMF8
+    : public MultiHeadedAttentionTest {
+ protected:
+  const std::string  // NOLINT
+  GetModuleFlash_Attention_Inference_BMM1_NoMask_Generation_Softmax_BMM2_HloString_BF16() {  // NOLINT
```
Comment (philipphack) [inline xla/service/gpu/tests/gpu_fused_mha_test.cc:1096]: Does this have to be a function? The name has 86 characters.

File: xla/service/gpu/tests/gpu_fused_mha_test.cc [hunk 17]
```diff
@@ -1089,6 +1089,119 @@ class FlashAttentionBMMScalePaddingMaskSoftmaxBMM
...
+      ROOT get-tuple-element.5.0 = bf16[4,4,16,16]{3,1,2,0} get-tuple-element(custom-call.4.0), index=0, metadata={op_name="jit(<unnamed wrapped function>)/jit(main)/dot_product_attention_fwd[scale=1.0 seed=42 dropout_rate=0.0 variadic_args=(False, False, False) mask_type=MaskType.NO_MASK layout=0 is_training=False]" source_file="/opt/rosetta/test_num.py" source_line=331}
+    } // main.106
+  )";
+    return hlo_text;
+  }
+
+  const std::string  // NOLINT
+  GetModuleFlash_Attention_Inference_BMM1_NoMask_Generation_Softmax_BMM2_HloString_FP8() {  // NOLINT
```
Comment (philipphack) [inline xla/service/gpu/tests/gpu_fused_mha_test.cc:1139]: Same as above.

File: xla/service/gpu/tests/gpu_fused_mha_test.cc [hunk 18]
```diff
@@ -1089,6 +1089,119 @@ class FlashAttentionBMMScalePaddingMaskSoftmaxBMM
...
+      get-tuple-element.5.0 = f8e4m3fn[4,4,16,16]{3,1,2,0} get-tuple-element(custom-call.21.0), index=0, metadata={op_name="jit(<unnamed wrapped function>)/jit(main)/dot_product_attention_fwd[scale=1.0 seed=42 dropout_rate=0.0 variadic_args=(False, False, False) mask_type=MaskType.NO_MASK layout=0 is_training=False]" source_file="/opt/rosetta/test_num.py" source_line=331}
+      ROOT out = bf16[4,4,16,16]{3,1,2,0} convert(get-tuple-element.5.0)
+    } // main.106
+  )";
+    return hlo_text;
+  }
+
+  template <typename T>
```
Comment (philipphack) [inline xla/service/gpu/tests/gpu_fused_mha_test.cc:1184]: Where is T used?

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 19]
```diff
@@ -8410,6 +8607,8 @@ absl::Status CudnnGraph::Execute(Stream& stream,
   std::unordered_map<int64_t, void*> tensor_to_ptr_map;
   absl::Span<DeviceMemoryBase> operands_without_workspace = operands;
   DeviceMemoryBase workspace;
+  std::cout << "Cudnngraph::Execute with wks=" << graph_.get_workspace_size()
```
Comment (philipphack) [inline xla/stream_executor/cuda/cuda_dnn.cc:8610]: Remove or replace with VLOG.

**Review by philipphack — CHANGES_REQUESTED** [review]

File: xla/service/gpu/cublas_cudnn.cc (hunk 1 above)
Comment (wenscarl) [reply xla/service/gpu/cublas_cudnn.cc:79]: I follow the convention in "__cublas$lt$matmul$f8"; so there is a dollar sign.

File: xla/service/gpu/cudnn_workspace_rewriter.cc (hunk 4 above)
Comment (wenscarl) [reply xla/service/gpu/cudnn_workspace_rewriter.cc:238]: The outputs of fp8 call are: O, amax_s, amax_o and activation(attn score).

File: xla/service/gpu/gpu_fused_mha_runner.cc [hunk 20]
```diff
@@ -272,6 +349,84 @@ absl::Status RunGpuFMHABackwardImpl(const GpufMHABackwardParams &params,
...
+/*static*/ absl::StatusOr<GpufMHAConfig> GpufMHAConfig::For(
+    const GpufMHAF8Descriptor &desc) {
+  // Get shapes from desc.
+  const Shape &lhs_bmm1_shape = desc.lhs_bmm1_shape;
+  const Shape &rhs_bmm1_shape = desc.rhs_bmm1_shape;
+  const Shape &rhs_bmm2_shape = desc.rhs_bmm2_shape;
+  const Shape &intermediate_lhs_bmm2_shape = desc.intermediate_lhs_bmm2_shape;
+  const Shape &output_shape = desc.output_shapes[0];
+
+  // Get DNN dtype from primtive types
```
Comment (akuegel) [inline xla/service/gpu/gpu_fused_mha_runner.cc:362]: nit: primtive -> primitive

File: xla/service/gpu/gpu_fused_mha_runner.cc [hunk 21]
```diff
@@ -272,6 +349,84 @@ absl::Status RunGpuFMHABackwardImpl(const GpufMHABackwardParams &params,
...
+      desc.bmm2_dnums.lhs_batch_dimensions(),
+      desc.bmm2_dnums.lhs_contracting_dimensions());
+
+  config.output = TensorDescriptor::For(output_type, output_shape.dimensions(),
+                                        output_shape.layout().minor_to_major());
+
+  if (desc.output_shapes.size() > 3) {
+    const Shape &activation_shape = desc.output_shapes.back();
+    // Generally, activation should have same type as output, but set it
+    // explicityly just to be safe.
```
Comment (akuegel) [inline xla/service/gpu/gpu_fused_mha_runner.cc:413]: nit: explicityly -> explicitly

File: xla/service/gpu/gpu_fused_mha_runner.cc [hunk 22]
```diff
@@ -524,6 +679,20 @@ GpufMHAConfig::AsDnnFusedMHAOpConfig() const {
   return config;
 }
 
+absl::StatusOr<se::dnn::FusedMHAF8Op::Config>
+GpufMHAConfig::AsDnnFusedMHAF8OpConfig() const {
+  double scale = 1.0;
```
Comment (akuegel) [inline xla/service/gpu/gpu_fused_mha_runner.cc:684]: simpler: double scale = fmha_scale.value_or(1.0);

File: xla/service/gpu/tests/gpu_fused_mha_test.cc [hunk 23]
```diff
@@ -1089,6 +1089,101 @@ class FlashAttentionBMMScalePaddingMaskSoftmaxBMM
...
 };
 
+class FlashAttentionBMMScalePaddingMaskSoftmaxBMMF8
+    : public MultiHeadedAttentionTest {
+ protected:
+  void TestImpl_Flash_Attention_Inference_BMM1_NoMask_Softmax_BMM2_F8() {
+    if (skip_reason_) GTEST_SKIP() << *skip_reason_;
+    if (GetDnnVersionInfoOrDefault(backend().default_stream_executor()) <
+        se::dnn::VersionInfo(8, 9, 3)) {
+      GTEST_SKIP() << "Flash Attention requires cuDNN >= 8.9.3.";
```
Comment (akuegel) [inline xla/service/gpu/tests/gpu_fused_mha_test.cc:1099]: I think recently we switched to requiring 9.0.0. So I think you should update this here.

File: xla/service/gpu/tests/gpu_fused_mha_test.cc [hunk 24]
```diff
@@ -1089,6 +1089,101 @@ class FlashAttentionBMMScalePaddingMaskSoftmaxBMM
...
+    }
+    XlaBuilder builder(TestName());
+    std::string hlo_string_ref =
+        R"(
+    HloModule jit__unnamed_wrapped_function_, entry_computation_layout={(bf16[4,16,4,16]{3,2,1,0}, bf16[4,16,4,16]{3,2,1,0}, bf16[4,16,4,16]{3,2,1,0})->bf16[4,4,16,16]{3,1,2,0}}, allow_spmd_sharding_propagation_to_parameters={true,true,true}, allow_spmd_sharding_propagation_to_output={true}
```
Comment (akuegel) [inline xla/service/gpu/tests/gpu_fused_mha_test.cc:1104]: Can you please adjust this HLO to remove the metadata? This makes it harder to read the actually relevant parts.

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 25]
```diff
@@ -7195,6 +7202,196 @@ int64_t GetDropoutRngOffset(std::vector<int64_t>& intermediate_shape) {
...
+    dnn::DnnSupport& dnn_support,
+    const dnn::MatmulTensorDescriptor& q_descriptor,
+    const dnn::MatmulTensorDescriptor& k_descriptor,
+    const dnn::MatmulTensorDescriptor& v_descriptor,
+    const dnn::TensorDescriptor& o_descriptor,
+    const std::optional<dnn::TensorDescriptor> stats_descriptor,
+    const float scale, const dnn::FMHAMaskKind mask_type) {
+  using cudnn_frontend::graph::Tensor_attributes;
+
+#if CUDNN_VERSION >= 8904
```
Comment (akuegel) [inline xla/stream_executor/cuda/cuda_dnn.cc:7215]: Same here, I think by now we want to require >= 90000

@Cjkkkk to confirm.

**Review by akuegel — CHANGES_REQUESTED** [review]

**Review by akuegel — APPROVED** [review]

**Comment by wenscarl** [comment]
Rebase and restructure based on #15919 .

**Review by akuegel — APPROVED** [review]

File: xla/service/gpu/tests/gpu_fused_mha_test.cc [hunk 26]
```diff
@@ -1150,6 +1239,12 @@ XLA_TEST_F(FlashAttentionBMMScalePaddingMaskSoftmaxBMM,
            Flash_Attention_Training_BMM1_PaddingMask_Softmax_BMM2_BF16) {
   TestImpl_Flash_Attention_Training_BMM1_PaddingMask_Softmax_BMM2<bfloat16>();
 }
+
+// BMM1 - Scale - Softmax - BMM2 fp8
+XLA_TEST_F(FlashAttentionBMMScalePaddingMaskSoftmaxBMMF8,
+           Flash_Attention_Inference_BMM1_NoMask_Softmax_BMM2_F8) {
```
Comment (akuegel) [inline xla/service/gpu/tests/gpu_fused_mha_test.cc:1245]: It seems this test is failing internally with

UNIMPLEMENTED: No registered implementation for custom call to __cudnn$fmhaSoftmax$f8 for platform CUDA

Is there anything else required than cudnn 9.0 or above?

**Review by akuegel — APPROVED** [review]

**Comment by penpornk** [comment]
Thank you for the fix! Now we get a driver version error:
```
[ RUN      ] FlashAttentionBMMScalePaddingMaskSoftmaxBMMF8.Flash_Attention_Inference_BMM1_NoMask_Softmax_BMM2_F8
third_party/tensorflow/compiler/xla/service/gpu/tests/gpu_fused_mha_test.cc:1190: Failure
Value of: RunAndCompareTwoModules(hlo_string, hlo_string_ref, ErrorSpec{1e-2, 1e-2})
  Actual: false (INTERNAL: sdpa fp8 forward operation is only supported starting cudnn 9.1.0. Please consider upgrading your current version.
in third_party/tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc(8226): 'graph_.validate()' 
```
Could you please add a version check?

**Review by wenscarl — COMMENTED** [review]
Update version check to v9.1.0

**Review by akuegel — APPROVED** [review]
